README.md

![logo](https://github.com/benmaier/epipack/raw/master/img/logo_12_lila_medium.png)

Fast prototyping of epidemiological models based on reaction equations.
Analyze the ODEs analytically or numerically, or run/animate stochastic 
simulations on networks/well-mixed systems.

* repository: https://github.com/benmaier/epipack/
* documentation: http://epipack.benmaier.org/

```python
import epipack as epk
from epipack.vis import visualize
import netwulf as nw

network, _, __ = nw.load('cookbook/readme_vis/MHRN.json')
N = len(network['nodes'])
links = [ (l['source'], l['target'], 1.0) for l in network['links'] ]

S, I, R = list("SIR")
model = epk.StochasticEpiModel([S,I,R],N,links)\
            .set_link_transmission_processes([ (I, S, 1.0, I, I) ])\
            .set_node_transition_processes([ (I, 1.0, R) ])\
            .set_random_initial_conditions({ S: N-5, I: 5 })

visualize(model, network, sampling_dt=0.1)
```

![sir-example](https://github.com/benmaier/epipack/raw/master/img/SIR_example.gif)

## Idea

Simple compartmental models of infectious diseases are useful
to investigate effects of certain processes on disease dissemination.
Using pen and paper, quickly adding/removing compartments and transition processes
is easy, yet the analytical and numerical analysis or stochastic simulations
can be tedious to set up and debug—especially when the model changes (even slightly).
`epipack` aims at streamlining this process
such that all the analysis steps can be performed in an efficient manner,
simply by defining processes based on reaction equations. `epipack` provides
three main base classes to accomodate different problems.

* `EpiModel`: Define a model based on transition, birth, 
  death, fission, fusion, or transmission reactions, integrate the 
  ordinary differential equations (ODEs) of the corresponding well-mixed system
  numerically or simulate it using Gillespie's algorithm.
  Process rates can be numerical functions of time and the system state.
* `SymbolicEpiModel`: Define a model based on transition, birth, 
  death, fission, fusion, or transmission reactions. Obtain the ODEs,
  fixed points, Jacobian, and the Jacobian's eigenvalues at fixed points
  as symbolic expressions. 
  Process rates can be symbolic expressions of time and the system state.
  Set numerical parameter values and integrate the ODEs numerically or
  simulate the stochastic systems using Gillespie's algorithm.
* `StochasticEpiModel`: Define a model based on node transition and
  link transmission reactions. Add conditional link transmission reactions.
  Simulate your model on any (un-/)directed, (un-/)weighted static/temporal
  network, or in a well-mixed system.

Additionally, epipack provides a visualization framework to animate
stochastic simulations on networks, lattices, well-mixed systems,
or reaction-diffusion systems based on `MatrixEpiModel`.

Check out the [Example](#examples) section for some demos.

Note that the internal simulation algorithm for network simulations is
based on the following paper:

"Efficient sampling of spreading processes on complex networks 
using a composition and rejection algorithm", G.St-Onge, 
J.-G. Young, L. Hébert-Dufresne, and L. J. Dubé, Comput.
Phys. Commun. 240, 30-37 (2019), http://arxiv.org/abs/1808.05859.

## Install

    pip install epipack

`epipack` was developed and tested for 

* Python 3.6
* Python 3.7
* Python 3.8

So far, the package's functionality was tested on Mac OS X and CentOS only.

## Dependencies

`epipack` directly depends on the following packages which will be installed by `pip` during the installation process

* `numpy>=1.17`
* `scipy>=1.3`
* `sympy==1.6`
* `pyglet<1.6`
* `matplotlib>=3.0.0`
* `ipython>=7.14.0`
* `ipywidgets>=7.5.1`

Please note that **fast network simulations are only available if you install** 

* `SamplableSet==2.0` ([SamplableSet](http://github.com/gstonge/SamplableSet))

**manually** (pip won't do it for you).

## Documentation

The full documentation is available at [epipack.benmaier.org](http://epipack.benmaier.org).

## Changelog

Changes are logged in a [separate file](https://github.com/benmaier/epipack/blob/master/CHANGELOG.md).

## License

This project is licensed under the [MIT License](https://github.com/benmaier/epipack/blob/master/LICENSE).
Note that this excludes any images/pictures/figures shown here or in the documentation.

## Contributing

If you want to contribute to this project, please make sure to read the [code of conduct](https://github.com/benmaier/epipack/blob/master/CODE_OF_CONDUCT.md) and the [contributing guidelines](https://github.com/benmaier/epipack/blob/master/CONTRIBUTING.md). In case you're wondering about what to contribute, we're always collecting ideas of what we want to implement next in the [outlook notes](https://github.com/benmaier/epipack/blob/master/OUTLOOK.md).

[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-v1.4%20adopted-ff69b4.svg)](code-of-conduct.md)

## Examples

Let's define an SIRS model with infection rate `eta`, recovery rate `rho`, and waning immunity rate `omega` and analyze the system

### Pure Numeric Models

#### Basic Definition (EpiModel)

Define a pure numeric model with `EpiModel`.
Integrate the ODEs or simulate the system stochastically.

```python
from epipack import EpiModel
import matplotlib.pyplot as plt
import numpy as np

S, I, R = list("SIR")
N = 1000

SIRS = EpiModel([S,I,R],N)\
    .set_processes([
        #### transmission process ####
        # S + I (eta=2.5/d)-> I + I
        (S, I, 2.5, I, I),

        #### transition processes ####
        # I (rho=1/d)-> R
        # R (omega=1/14d)-> S
        (I, 1, R),
        (R, 1/14, S),
    ])\
    .set_initial_conditions({S:N-10, I:10})

t = np.linspace(0,40,1000) 
result_int = SIRS.integrate(t)
t_sim, result_sim = SIRS.simulate(t[-1])

for C in SIRS.compartments:
    plt.plot(t, result_int[C])
    plt.plot(t_sim, result_sim[C])
```

![numeric-model](https://github.com/benmaier/epipack/raw/master/img/numeric_model.png)

#### Functional Rates

It's also straight-forward to define temporally varying (functional) rates.

```python
import numpy as np
from epipack import SISModel

N = 100
recovery_rate = 1.0

def infection_rate(t, y, *args, **kwargs):
    return 3 + np.sin(2*np.pi*t/100)

SIS = SISModel(
            infection_rate=infection_rate, 
            recovery_rate=recovery_rate,
            initial_population_size=N
            )\
        .set_initial_conditions({
            'S': 90,
            'I': 10,
        })

t = np.arange(200)
result_int = SIS.integrate(t)
t_sim, result_sim = SIS.simulate(199)

for C in SIS.compartments:
    plt.plot(t_sim, result_sim[C])
    plt.plot(t, result_int[C])
```

![numeric-model-time-varying](https://github.com/benmaier/epipack/raw/master/img/numeric_model_time_varying_rate.png)

### Symbolic Models

#### Basic Definition

Symbolic models are more powerful because they can 
do the same as the pure numeric models while 
also offering the possibility to do analytical evaluations

```python
from epipack import SymbolicEpiModel
import sympy as sy

S, I, R, eta, rho, omega = sy.symbols("S I R eta rho omega")

SIRS = SymbolicEpiModel([S,I,R])\
    .set_processes([
        (S, I, eta, I, I),
        (I, rho, R),
        (R, omega, S),
    ])    
```

#### Analytical Evaluations

Print the ODE system in a Jupyter notebook

```python
>>> SIRS.ODEs_jupyter()
```

![ODEs](https://github.com/benmaier/epipack/raw/master/img/ODEs.png)

Get the Jacobian

```python
>>> SIRS.jacobian()
```

![Jacobian](https://github.com/benmaier/epipack/raw/master/img/jacobian.png)

Find the fixed points

```python
>>> SIRS.find_fixed_points()
```

![fixedpoints](https://github.com/benmaier/epipack/raw/master/img/fixed_points.png)

Get the eigenvalues at the disease-free state in order to find the epidemic threshold

```python
>>> SIRS.get_eigenvalues_at_disease_free_state()
{-omega: 1, eta - rho: 1, 0: 1}
```

#### Numerical Evaluations

Set numerical parameter values and
integrate the ODEs numerically

```python
>>> SIRS.set_parameter_values({eta: 2.5, rho: 1.0, omega:1/14})
>>> t = np.linspace(0,40,1000)
>>> result = SIRS.integrate(t)
```

If set up as

```python
>>> N = 10000
>>> SIRS = SymbolicEpiModel([S,I,R],N)
```

the system can simulated directly.

```python
>>> t_sim, result_sim = SIRS.simulate(40)
```

#### Temporally Varying Rates

Let's set up some temporally varying rates

```python
from epipack import SymbolicEpiModel
import sympy as sy

S, I, R, eta, rho, omega, t, T = \
        sy.symbols("S I R eta rho omega t T")

N = 1000
SIRS = SymbolicEpiModel([S,I,R],N)\
    .set_processes([
        (S, I, 2+sy.cos(2*sy.pi*t/T), I, I),
        (I, rho, R),
        (R, omega, S),
    ])  

SIRS.ODEs_jupyter()
```


![SIRS-forced-ODEs](https://github.com/benmaier/epipack/raw/master/img/SIRS-forced-ODEs.png)

Now we can integrate the ODEs or
simulate the system using Gillespie's SSA
for inhomogeneous Poisson processes.

```python
import numpy as np

SIRS.set_parameter_values({
    rho : 1,
    omega : 1/14,
    T : 100,
})
SIRS.set_initial_conditions({S:N-100, I:100})
_t = np.linspace(0,200,1000)
result = SIRS.integrate(_t)
t_sim, result_sim = SIRS.simulate(max(_t))
```

![SIRS-forced-results](https://github.com/benmaier/epipack/raw/master/img/symbolic_model_time_varying_rate.png)

#### Interactive Analyses

`epipack` offers a classs called `InteractiveIntegrator`
that allows an interactive exploration of a system 
in a Jupyter notebook.

Make sure to first run

```
%matplotlib widget
```

in a cell.

```python
from epipack import SymbolicEpiModel
from epipack.interactive import InteractiveIntegrator, Range, LogRange
import sympy

S, I, R, R0, tau, omega = sympy.symbols("S I R R_0 tau omega")

I0 = 0.01
model = SymbolicEpiModel([S,I,R])\
             .set_processes([
                    (S, I, R0/tau, I, I),
                    (I, 1/tau, R),
                    (R, omega, S),
                ])\
             .set_initial_conditions({S:1-I0, I:I0})

# define a log slider, a linear slider and a constant value
parameters = {
    R0: LogRange(min=0.1,max=10,step_count=1000),
    tau: Range(min=0.1,max=10,value=8.0),
    omega: 1/14
}

t = np.logspace(-3,2,1000)
InteractiveIntegrator(model, parameters, t, figsize=(4,4))
```

![interactive](https://github.com/benmaier/epipack/raw/master/img/interactive.gif)

### Pure Stochastic Models

#### On a Network

Let's simulate an SIRS system on a random graph
(using the parameter definitions above).

```python
from epipack import StochasticEpiModel
import networkx as nx

k0 = 50
R0 = 2.5
rho = 1
eta = R0 * rho / k0
omega = 1/14
N = int(1e4)
edges = [ (e[0], e[1], 1.0) for e in \
          nx.fast_gnp_random_graph(N,k0/(N-1)).edges() ]

SIRS = StochasticEpiModel(
            compartments=list('SIR'),
            N=N,
            edge_weight_tuples=edges
            )\
        .set_link_transmission_processes([
            ('I', 'S', eta, 'I', 'I'),
        ])\
        .set_node_transition_processes([
            ('I', rho, 'R'),
            ('R', omega, 'S'),
        ])\        
        .set_random_initial_conditions({
                                        'S': N-100,
                                        'I': 100
                                       })
t_s, result_s = SIRS.simulate(40)
```

![network-simulation](https://github.com/benmaier/epipack/raw/master/img/network_simulation.png)

#### Visualize

Likewise, it's straight-forward to visualize this system

```python
>>> from epipack.vis import visualize
>>> from epipack.networks import get_random_layout
>>> layouted_network = get_random_layout(N, edges)
>>> visualize(SIRS, layouted_network, sampling_dt=0.1, config={'draw_links': False})
```

![sirs-example](https://github.com/benmaier/epipack/raw/master/img/SIRS_visualization.gif)


#### On a Lattice

A lattice is nothing but a network, we can use
`get_grid_layout` and `get_2D_lattice_links`
to set up a visualization. 

```python
from epipack.vis import visualize
from epipack import (
    StochasticSIRModel, 
    get_2D_lattice_links, 
    get_grid_layout
)

# define links and network layout
N_side = 100
N = N_side**2
links = get_2D_lattice_links(N_side, periodic=True, diagonal_links=True)
lattice = get_grid_layout(N)

# define model
R0 = 3; recovery_rate = 1/8
model = StochasticSIRModel(N,R0,recovery_rate,
                           edge_weight_tuples=links)
model.set_random_initial_conditions({'I':20,'S':N-20})

sampling_dt = 1

visualize(model,lattice,sampling_dt,
        config={
                 'draw_nodes_as_rectangles':True,
                 'draw_links':False,
               }
          )
```

![sir-lattice](https://github.com/benmaier/epipack/raw/master/img/SIR_lattice_vis.gif)

### Reaction-Diffusion Models

Since reaction-diffusion systems in discrete space
can be interpreted as being based on reaction
equations, we can set those up using `epipack`'s
framework.  

Checkout the docs on
[Reaction-Diffusion Systems](http://epipack.benmaier.org/tutorial/reaction_diffusion.html).

Every node in a network is associated
with a compartment and we're using `MatrixEpiModel`
because it's faster than `EpiModel`.

```python
from epipack import MatrixEpiModel

N = 100
base_compartments = list("SIR")
compartments = [ (node, C) for node in range(N) for C in base_compartments ]
model = MatrixEpiModel(compartments)
```

Now, we define both epidemiological
and movement processes on a hypothetical
list `links`.

```python
infection_rate = 2
recovery_rate = 1
mobility_rate = 0.1

quadratic_processes = []
linear_processes = []

for node in range(N):
    quadratic_processes.append(
            (  (node, "S"), (node, "I"), infection_rate, (node, "I"), (node, "I") ),
        )

    linear_processes.append(
              ( (node, "I"), recovery_rate, (node, "R") ) 
        )

for u, v, w in links:
    for C in base_compartments:

        linear_processes.extend([
                  ( (u, C), w*mobility_rate, (v, C) ),
                  ( (v, C), w*mobility_rate, (u, C) ),
            ])
```

![reac-diff-lattice](https://github.com/benmaier/epipack/raw/master/img/reac_diff_lattice.gif)

## Dev notes

Fork this repository, clone it, and install it in dev mode.

```bash
git clone git@github.com:YOURUSERNAME/epipack.git
make
```

If you want to upload to PyPI, first convert the new `README.md` to `README.rst`

```bash
make readme
```

It will give you warnings about bad `.rst`-syntax. Fix those errors in `README.rst`. Then wrap the whole thing 

```bash
make pypi
```

It will probably give you more warnings about `.rst`-syntax. Fix those until the warnings disappear. Then do

```bash
make upload
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## [v0.1.7] - 2021-10-18

### Added

- it's possible to set explicit time dependence of rates now when starting a simulation

### Fixed

- changed references to deprecated classes in examples and docstrings
- better handling of detecting explicit time-dependence in functional rates.

## [v0.1.6] - 2021-07-30

### Changed

- renamed a parameter in `StochasticEpiModel.simulate` to `stop_simulation_on_vanishing_total_event_rate`, because it describes the mechanism better (adjusted `TemporalNetwork`) accordingly

### Fixed

- Disallowing resizing a visualization app on Linux OSes to avoid a recursion error
- metadata such as `epipack.__version__` is now correctly given out

### Added

- a `SymbolicSEIRModel`
- a routine to generate random geometric graphs
- the possiblity to add a callback function in `StochasticEpiModel.simulate` such that event tracking is possible

## [v0.1.5] - 2021-06-08

### Added

- methods to the IntegratorMixin class that allow integration until a stop condition is reached

### Fixed

- specified for MatrixEpiModel that the leading eigenvalue of a Jacobian should be measured by largest real part
- specified for MatrixEpiModel that the leading eigenvalue of the next generation matrix should be measured by largest magnitude (because R is the matrix's spectral radius)
- in StochasticEpiModel, convert rate values to float by default

## [v0.1.4] - 2021-05-18

### Added

- A new small-world network styling method based on 1d lattice distance (in epipack.networks)
- methods to compute Jacobian and next generation matrices (NGMs) in `MatrixEpiModel`, as well as R0 from said NGMs (TODO: add docs)
- tests for these methods
- `epipack.distributions` module, which deals with fitting empirical distributions to sums of exponentially distributed random variables (still in dev mode, also TODO: add docs)
- tests for this module
- methods to `EpiModel` that save events that have been set. This will be used to generate model flowcharts with graphviz at some point
- the possibility to pass a function to ``StochasticEpiModel.simulate`` that checks for a custom stop condition

## [v0.1.3] - 2021-04-07

### Fixed

- dependency issues with pyglet, apparently the "shapes" module did not appear until lately. Defined a range of versions for pyglet
- bug in example code in README.md

## [v0.1.2] - 2021-04-01

### Fixed

- A bug where the `reset_events`-flag was ignored when setting processes

## [v0.1.1] - 2021-03-03

### Added

- GeneralInteractiveWidget was added to allow interactive display of general functions
- a very basic SDE integrator was added (no diffusion coefficent matrix, and no system-dependent diffusion coefficients)

### Changed

- InteractiveIntegrator can now plot derivatives
- Range and LogRange classes will behave like floats whenever necessary

### Fixed

- behavior of the SamplableSet class
- a bug where the reaction rate of nodes in weighted networks is scaled by the node's degree and not by its strength

## [v0.1.0] - 2020-10-21

### Added
- `epipack.interactive`: contains a class that adds an interactive widget to Jupyter notebooks
  with which one may control the parameter values of a SymbolicEpiModel instance
- `epipack.temporal_networks`: set up temporal networks and model simulations on them
- `SymbolicODEModel`: A model that's defined via ODEs in sympy format.

## [v0.0.5] - 2020-08-14

### Changed
- DeterministicEpiModel is now MatrixEpiModel
- SymbolicEpiModel is now SymbolicMatrixEpiModel
- in StochasticEpiModel and during visualization, a more efficient mechanism checks for whether the simulation has ended for good

### Added
- Added models that are based entirely on events. In this way, we can easily implement time-dependent rates and have single models that can do everything at once: symbolic evaluations, numerical evaluations, and stochastic mean-field simulations
- time-dependent rates are integrated using a time-varying Gillespie algorithm
- EpiModel, StochasticSIModel, StochasticSIRModel, StochasticSISModel, SymbolicEpiModel (based on events rather than rates)

## [v0.0.4] - 2020-08-03
### Changed
- SymbolicEpiModel: raise error when `disease_free_state` is not given explicitly and no S-compartment can be found
- allow non-unity initial conditions for SymbolicEpiModel and DeterministicEpiModel
- `population_size` is now explicitly regarded in SymbolicEpiModel
- in DeterministicEpiModel, instead of raising errors, warnings are raised for nonzero column sums
- in StochasticEpiModel, save the current state after the end of the simulation

### Added
- A complete visualization framework and network grid layout
- in StochasticEpiModel, a callback function can be passed that's called whenever a sample is taken during the simimulation

### Fixed
- fixed bug where fission processes were converted to quadratic rates

## [v0.0.3] - 2020-06-30
### Added
- Catch situations where the true total event is zero but the maximum total event rate is non-zero

## [v0.0.2] - 2020-06-29
### Changed
- Catching ModuleNotFoundError properly

## [v0.0.1] - 2020-06-25
### Added
- Working package

## v0.0.0 - 2020-06-22
### Changed
- initialized

[Unreleased]: https://github.com/benmaier/epipack/compare/v0.1.7...HEAD
[v0.1.7]: https://github.com/benmaier/epipack/compare/v0.1.6...v0.1.7]
[v0.1.6]: https://github.com/benmaier/epipack/compare/v0.1.5...v0.1.6]
[v0.1.5]: https://github.com/benmaier/epipack/compare/v0.1.4...v0.1.5]
[v0.1.4]: https://github.com/benmaier/epipack/compare/v0.1.3...v0.1.4]
[v0.1.3]: https://github.com/benmaier/epipack/compare/v0.1.2...v0.1.3]
[v0.1.2]: https://github.com/benmaier/epipack/compare/v0.1.1...v0.1.2]
[v0.1.1]: https://github.com/benmaier/epipack/compare/v0.1.0...v0.1.1]

img/README_plots.py

def network_simulation():
    from epipack import StochasticEpiModel
    from epipack.plottools import plot
    import matplotlib.pyplot as pl
    import networkx as nx

    k0 = 50
    R0 = 2.5
    rho = 1
    eta = R0 * rho / k0
    omega = 1/14
    N = int(1e4)
    edges = [ (e[0], e[1], 1.0) for e in \
              nx.fast_gnp_random_graph(N,k0/(N-1)).edges() ]

    SIRS = StochasticEpiModel(
                compartments=list('SIR'),
                N=N,
                edge_weight_tuples=edges
                )\
            .set_link_transmission_processes([
                ('I', 'S', eta, 'I', 'I'),
            ])\
            .set_node_transition_processes([
                ('I', rho, 'R'),
                ('R', omega, 'S'),
            ])\
            .set_random_initial_conditions({
                                            'S': N-100,
                                            'I': 100
                                           })
    t_s, result_s = SIRS.simulate(40)

    ax = plot(t_s, result_s)
    ax.get_figure().savefig('network_simulation.png',dpi=300)


def network_simulation_visualization():
    from epipack import StochasticEpiModel
    from epipack.plottools import plot
    import matplotlib.pyplot as pl
    import networkx as nx

    k0 = 50
    R0 = 2.5
    rho = 1
    eta = R0 * rho / k0
    omega = 1/14
    N = int(1e4)
    edges = [ (e[0], e[1], 1.0) for e in \
              nx.fast_gnp_random_graph(N,k0/(N-1)).edges() ]

    SIRS = StochasticEpiModel(
                compartments=list('SIR'),
                N=N,
                edge_weight_tuples=edges
                )\
            .set_link_transmission_processes([
                ('I', 'S', eta, 'I', 'I'),
            ])\
            .set_node_transition_processes([
                ('I', rho, 'R'),
                ('R', omega, 'S'),
            ])\
            .set_random_initial_conditions({
                                            'S': N-100,
                                            'I': 100
                                           })

    from epipack.vis import visualize
    from epipack.networks import get_random_layout
    layouted_network = get_random_layout(N, edges)
    visualize(SIRS, layouted_network, sampling_dt=0.1, config={'draw_links': False})

def MHRN_network_vis():
    import epipack as epk
    from epipack.vis import visualize
    import netwulf as nw

    network, _, __ = nw.load('../cookbook/readme_vis/MHRN.json')
    N = len(network['nodes'])
    links = [ (l['source'], l['target'], 1.0) for l in network['links'] ]

    model = epk.StochasticEpiModel(["S","I","R"],N,links)\
                .set_link_transmission_processes([ ("I", "S", 1.0, "I", "I") ])\
                .set_node_transition_processes([ ("I", 1.0, "R") ])\
                .set_random_initial_conditions({ "S": N-5, "I": 5 })

    visualize(model, network, sampling_dt=0.1)

def default_numeric_models():
    from epipack import EpiModel
    import matplotlib.pyplot as pl
    from epipack.plottools import plot
    import numpy as np

    S, I, R = list("SIR")
    N = 1000

    SIRS = EpiModel([S,I,R],N)\
        .set_processes([
            #### transmission process ####
            # S + I (eta=2.5/d)-> I + I
            (S, I, 2.5, I, I),

            #### transition processes ####
            # I (rho=1/d)-> R
            # R (omega=1/14d)-> S
            (I, 1, R),
            (R, 1/14, S),
        ])\
        .set_initial_conditions({S:N-10, I:10})

    t = np.linspace(0,40,1000)
    result_int = SIRS.integrate(t)
    t_sim, result_sim = SIRS.simulate(t[-1])

    ax = plot(t_sim, result_sim)
    ax = plot(t, result_int,ax=ax)
    ax.get_figure().savefig('numeric_model.png',dpi=300)

def varying_rate_numeric_models():

    import numpy as np
    from epipack import SISModel
    from epipack.plottools import plot

    N = 100
    recovery_rate = 1.0

    def infection_rate(t, y, *args, **kwargs):
        return 3 + np.sin(2*np.pi*t/100)

    SIS = SISModel(
                infection_rate=infection_rate,
                recovery_rate=recovery_rate,
                initial_population_size=N
                )\
            .set_initial_conditions({
                'S': 90,
                'I': 10,
            })

    t = np.arange(200)
    result_int = SIS.integrate(t)
    t_sim, result_sim = SIS.simulate(199)

    ax = plot(t_sim, result_sim)
    ax = plot(t, result_int,ax=ax)
    ax.get_figure().savefig('numeric_model_time_varying_rate.png',dpi=300)

def symbolic_simulation_temporally_forced():
    from epipack import SymbolicEpiModel
    import sympy as sy
    from epipack.plottools import plot
    import numpy as np

    S, I, R, eta, rho, omega, t, T = \
            sy.symbols("S I R eta rho omega t T")

    N = 1000
    SIRS = SymbolicEpiModel([S,I,R],N)\
        .set_processes([
            (S, I, 3+sy.cos(2*sy.pi*t/T), I, I),
            (I, rho, R),
            (R, omega, S),
        ])

    SIRS.set_parameter_values({
        rho : 1,
        omega : 1/14,
        T : 100,
    })
    SIRS.set_initial_conditions({S:N-100, I:100})
    _t = np.linspace(0,150,1000)
    result = SIRS.integrate(_t)
    t_sim, result_sim = SIRS.simulate(max(_t))

    ax = plot(_t, result)
    plot(t_sim, result_sim, ax=ax)
    ax.get_figure().savefig('symbolic_model_time_varying_rate.png',dpi=300)


if  __name__=="__main__":
    default_numeric_models()
    varying_rate_numeric_models()
    symbolic_simulation_temporally_forced()
    network_simulation()
    #network_simulation_visualization()
